### ISSUE TITLE
Bugsnag\Client's "notify" function has bug for callbacks

### ISSUE THREAD

**Issue by calvinps** [bugsnag/bugsnag-laravel#310]
### Expected behavior
$resolvedReport should not be null after calling the provided callback function from \Bugsnag::notifyException

### Observed behavior
$resolvedBehavior is an outside variable that you're trying to be set from inside a closure, which will not work, resulting in it being null when it should be set to the value of $report in the closure

### Additional information
This code works. look at the comment inside the callback. 

```
public function notify(Report $report, callable $callback = null)
 {$this->pipeline->execute($report, function ($report) use ($callback) {
            if ($callback) {
                $resolvedReport = null;
                $bridge = new CallbackBridge($callback);
                $bridge($report, function ($report) use ($resolvedReport) {
                    $resolvedReport = $report;
                    ### Added this line because $resolvedReport will be null outside this closure ###
                    $this->http->queue($resolvedReport);
                });
                if ($resolvedReport) {
                    $report = $resolvedReport;
                } else {
                    return;
                }
            }
            $this->http->queue($report);
        });

        $this->leaveBreadcrumb($report->getName(), Breadcrumb::ERROR_TYPE, $report->getSummary());
        if (!$this->config->isBatchSending()) {
            $this->flush();
        }
    }
```

**Comment by kattrali** [comment]
Hi @calvinps, thanks for the report. Can you add a snippet showing how the callback is being used?

**Comment by calvinps** [comment]
@kattrali Sure thing, see below. 

```
            \Bugsnag::notifyException($exc,function($report) use ($data){
                return $report->setMetaData([
                    'Test Data' => [
                        'trace' => $data,
                    ]
                ]);
            });
```

Here's some more info that might help too. 

When I include a callback function like the one above I observed that the bug report wasn't showing up on Bugsnag, but if I didn't include a callback it worked as expected. As you can see in my initial report, in the notify function of Bugsnag\Client there is an "if" statement checking to see if a callback is provided. At the end of the if clause if the $resolvedReport variable is null then the report is not sent to Bugsnag but rather the function returns without doing anything. Although $resolvedReport is being set within the callback, scope with anonymous functions in PHP works differently than you might expect - variables use the value from when they're defined, not when they're later used in callbacks. Thus, the changes in the callback to $resolvedReport don't apply outside of the closure (even if it appears above some other code).  This means that this section of code:

```
                if ($resolvedReport) {
                    $report = $resolvedReport;
                } else {
                    return;
                }
```
is run with the value of $resolvedReport being `null` (when it was declared) and not what it was later set to in the callback.  

Here's how I ended up changing the notify function to make it work locally. Not sure if this is the official solution you want to go with but hopefully it helps. 

```
    public function notify(Report $report, callable $callback = null)
    {
        $this->pipeline->execute($report, function ($report) use ($callback) {
            if ($callback) {
                $bridge = new CallbackBridge($callback);
                $bridge($report, function ($report) {
                    $this->http->queue($report);
                });
            } else {
                $this->http->queue($report);
            }

        });

        $this->leaveBreadcrumb($report->getName(), Breadcrumb::ERROR_TYPE, $report->getSummary());

        if (!$this->config->isBatchSending()) {
            $this->flush();
        }
    }
```

**Comment by pdbreen** [comment]
Why not just add & so assignment survives?  `$bridge($report, function ($report) use (&$resolvedReport)`

**Comment by calvinps** [comment]
Yep that works, much simpler fix!

**Comment by kattrali** [comment]
Thanks again! Released in bugsnag-php 3.14.1, run `composer update bugsnag/bugsnag` to update.
